flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java

```diff
@@ -41,18 +42,47 @@
   private final FileScanTaskReader<T> fileScanTaskReader;
 
   private final InputFilesDecryptor inputFilesDecryptor;
-  private Iterator<FileScanTask> tasks;
+  private final CombinedScanTask combinedTask;
+  private final Position position;
+
+  private Iterator<FileScanTask> fileTasksIterator;
   private CloseableIterator<T> currentIterator;
 
   public DataIterator(FileScanTaskReader<T> fileScanTaskReader, CombinedScanTask task,
                       FileIO io, EncryptionManager encryption) {
     this.fileScanTaskReader = fileScanTaskReader;
 
     this.inputFilesDecryptor = new InputFilesDecryptor(task, io, encryption);
-    this.tasks = task.files().iterator();
+    this.combinedTask = task;
+    // fileOffset starts at -1 because we started
+    // from an empty iterator that is not from the split files.
+    this.position = new Position(-1, 0L);
```

Comment:
The general `DataIterator` don't use the position or `seek` method to skip tasks or records.  Putting all the flip-27 related logics in the flink common read path does not make sense to me, because every times when I read this class, I need to see which part is related to flip-27, which is the unrelated part. 

I will suggest to introduce a separate SeekableDataIterator to isolate the two code path, I made a simple commit for this: https://github.com/openinx/incubator-iceberg/commit/b08dde86aae0c718d9d72acb347dffb3a836b336, you may want to take a look.

Reply:
I wouldn't say that `seek` capability is FLIP-27 specific. If we think `DataIterator` as reading a list of files/splits from `CombinedScanTask`, it is like a file API where `seek` is pretty common. It is needed to achieve exactly-once semantics on source reading. 

Thanks a lot for the `SeekableDataIterator`. I feel that leaving these two empty abstract methods in the base `DataIterator` is a little weird
```
protected void advanceRecord()
protected void advanceTask()
```

Overall, I still think adding `seek` capability to `DataIterator` is natural (for file-like read APIs)